### ISSUE TITLE
PINQ: TypeError: 'Table' object is not iterable

### ISSUE THREAD

**Issue by thedrow** [azazel75/macropy#1]
I'm trying to write a query using PINQ but I can't get it right for some reason.

```python
from macropy.experimental.pinq import *

statement = sql[((organization.id) for organization in organizations
                         if organization.enrollment_secret == enrollment_secret)]
```

raises the following exception:
```
TypeError: 'Table' object is not iterable
```

I checked and every one of your tests does the same thing and the tests pass.
Any suggestions are appreciated.

**Comment by azazel75** [comment]
Hi @thedrow,
that's strange!

Do you have any code that can be tested?

I have some suggestions from what i can see and my experience in debugging macros:

- try to narrow down the imports from `macros.experimental.pinq` to what you really use. If the macro doesn't work due to some mysterios missing symbol that's a bug that needs to be fixed;
- that parens around organization.id seem to do nothing to me, what you expect them to give you?
- I would try to activate the [SaveExporter for your module](http://macropy3.readthedocs.io/en/latest/export_code.html) to really see what code it generates and debug that.

**Comment by azazel75** [comment]
oh, forget my dumbness... You should signal the import machinery that your module has to be expanded

Try changing your imports to:
```python
from macropy.experimental.pinq import macros, *
```

**Comment by thedrow** [comment]
```
from macropy.experimental.pinq import macros, *
```
produces:
```
SyntaxError: invalid syntax
```

**Comment by thedrow** [comment]
I used the SaveExporter and it seems no code is expanded.

**Comment by thedrow** [comment]
In fact my code doesn't work for me even from the console with `import macropy.console`.

**Comment by azazel75** [comment]
Yes, indeed. See my last comment. You have to import `macros` explicitly as first name to activate macro expansion

**Comment by azazel75** [comment]
Sorry, I read your comment on import just now,
Please use 
```python
from macropy.experimental.pinq import macros, sql
```
and any other symbol you need

**Comment by thedrow** [comment]
Same result.
Here's the code:
```python
organizations = sa.Table(
    'organizations',
    metadata,
    sa.Column('id',
              UUID(as_uuid=True),
              primary_key=True,
              server_default=text("uuid_generate_v1mc()")),
    sa.Column('enrollment_secret',
              sa.String(255),
              nullable=False)
)
```

```python

from macropy.experimental.pinq import macros, sql
import secrets

from sqlalchemy.dialects import postgresql
from sqlalchemy.sql import select

from infraboost.databases import Repository
from infraboost.models import organizations

class OrganizationsRepository(Repository):
    class DoesNotExist(Exception):
        pass

    async def get_organization_id_by_enrollment_secret(self, enrollment_secret):
        # statement = select([organizations.columns.id]).where(
        #     organizations.columns.enrollment_secret == enrollment_secret)
        statement = sql[(organization.id for organization in organizations
                         if organization.enrollment_secret == enrollment_secret)]

        async with self._connection_pool.transaction() as conn:
            result = await conn.fetchval(statement)

        if result is None:
            raise OrganizationsRepository.DoesNotExist()

        return result

    async def create_organization(self):
        # TODO: Move the secret generation to the model
        enrollment_secret = secrets.token_hex()
        statement = postgresql.insert(organizations).values(
            enrollment_secret=enrollment_secret)

        async with self._connection_pool.transaction() as conn:
            await conn.execute(statement)

        return enrollment_secret
```

**Comment by azazel75** [comment]
Do you activate macropy?

**Comment by thedrow** [comment]
At one point I did import `macropy.activate` to use the save exporter as you asked but it showed no expansion. Let me try again.

**Comment by thedrow** [comment]
Ok here is the real error:
```
  File "/home/omer/Documents/Projects/infraboost/tests/functional/conftest.py", line 21, in <module>
    from infraboost.organizations.repositories import OrganizationsRepository
  File "/home/omer/Documents/Projects/infraboost/infraboost/organizations/repositories.py", line 2, in <module>
    import macropy.activate
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/activate.py", line 4, in <module>
    macropy.activate()
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/__init__.py", line 17, in activate
    from .core import failure
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 951, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 894, in _find_spec
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/import_hooks.py", line 135, in find_spec
    code, tree = self.expand_macros(source, origin, spec)
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/import_hooks.py", line 111, in expand_macros
    modules)
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/macros.py", line 251, in expand_entire_ast
    expand_macros=expand_macros, **file_vars)
TypeError: interned_name() missing 1 required positional argument: 'gen_sym'
```

**Comment by azazel75** [comment]
Can I ask you to install my master repo directly and repost any error? It is ready for release but I still need to fix some docs.
Use 
```console
pip install -U https://github.com/azazel75/macropy/archive/master.zip 
```

**Comment by thedrow** [comment]
```
  File "/home/omer/Documents/Projects/infraboost/tests/functional/conftest.py", line 21, in <module>
    from infraboost.organizations.repositories import OrganizationsRepository
  File "/home/omer/Documents/Projects/infraboost/infraboost/organizations/repositories.py", line 2, in <module>
    import macropy.activate
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/activate.py", line 4, in <module>
    macropy.activate()
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/__init__.py", line 19, in activate
    from .core import failure  # noqa
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 951, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 894, in _find_spec
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/import_hooks.py", line 139, in find_spec
    code, tree = self.expand_macros(source, origin, spec)
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/import_hooks.py", line 111, in expand_macros
    tree, source_code, modules).expand_macros()
  File "/home/omer/.local/share/virtualenvs/infraboost--RkZn49j/lib/python3.6/site-packages/macropy/core/macros.py", line 491, in __init__
    **self.file_vars)
TypeError: interned_name() missing 1 required positional argument: 'gen_sym'
```

**Comment by azazel75** [comment]
@thedrow would you be so kind to redo the installation from my master and re-running your code?

**Comment by azazel75** [comment]
It's very strange it shouldn't be needed... gen_sym is imported by macropy.activate

**Comment by azazel75** [comment]
which Py3.6 are you using?

**Comment by thedrow** [comment]
3.6.5

**Comment by thedrow** [comment]
I'll try again tomorrow.

**Comment by azazel75** [comment]
Ok,  so I'm a little bit confused here... I've actually tried your code in an empty dir. I've used the following files:

in `run.py`:
```python
import macropy
import macropy.logging
import macropy.activate
macropy.exporter = macropy.exporters.SaveExporter(
    'cmp', '.')

import omer
```
and then  in `omer.py` a slightly modified version just to let it import without issues (I don't have your model code):
```python
from macropy.experimental.pinq import macros, sql
import secrets

from sqlalchemy.dialects import postgresql
from sqlalchemy.sql import select

#from infraboost.databases import Repository
#from infraboost.models import organizations

class OrganizationsRepository:
    class DoesNotExist(Exception):
        pass

    async def get_organization_id_by_enrollment_secret(self, enrollment_secret):
        # statement = select([organizations.columns.id]).where(
        #     organizations.columns.enrollment_secret == enrollment_secret)
        statement = sql[(organization.id for organization in organizations
                         if organization.enrollment_secret == enrollment_secret)]

        async with self._connection_pool.transaction() as conn:
            result = await conn.fetchval(statement)

        if result is None:
            raise OrganizationsRepository.DoesNotExist()

        return result

    async def create_organization(self):
        # TODO: Move the secret generation to the model
        enrollment_secret = secrets.token_hex()
        statement = postgresql.insert(organizations).values(
            enrollment_secret=enrollment_secret)

        async with self._connection_pool.transaction() as conn:
            await conn.execute(statement)

        return enrollment_secret
```

after running `run.py` with Python 3.6.5, I found the following file `cmp/omer.py`:
```python
from pickle import _loads as unpickled
[sqlalchemy] = unpickled(b'\x80\x03]q\x00cmacropy.experimental.pinq\nsqlalchemy\nq\x01a.')
from macropy.experimental.pinq import macros
import secrets
from sqlalchemy.dialects import postgresql
from sqlalchemy.sql import select

class OrganizationsRepository:

    class DoesNotExist(Exception):
        pass

    async def get_organization_id_by_enrollment_secret(self, enrollment_secret):
        statement = (lambda organization: sqlalchemy.select([organization.id]).where((organization.enrollment_secret == enrollment_secret)))(organizations.alias().c)
        async with self._connection_pool.transaction() as conn:
            result = (await conn.fetchval(statement))
        if (result is None):
            raise OrganizationsRepository.DoesNotExist()
        return result

    async def create_organization(self):
        enrollment_secret = secrets.token_hex()
        statement = postgresql.insert(organizations).values(enrollment_secret=enrollment_secret)
        async with self._connection_pool.transaction() as conn:
            (await conn.execute(statement))
        return enrollment_secret
```

Which seems expanded as expected to me... all this without the patch that I linked here... so I think I'm going to rollback it unless you say to me that it matters.. It doesn't give problems here. Are you using something to make imports asynchronous? (sounds strange but...)

**Comment by thedrow** [comment]
The patch does make a difference. We're now back to `TypeError: 'Table' object is not iterable`.

**Comment by thedrow** [comment]
The secret is to use `import macropy.activate` in my main package's `__init__.py`.
I think that the patch you provided will prevent crashes when not doing so. Not sure how.

**Comment by azazel75** [comment]
So it works now?

**Comment by azazel75** [comment]
I'll rollback my last patch, I'm quite sure I was wrong with that

**Comment by thedrow** [comment]
Yes.
